pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,407 @@
+                new_tape, stop_at=~qml.operation.is_measurement, expand_measurements=True
+            )
+            num_tape += 1
+
+            g_tapes.append(new_tape)
+
+        gradient_data.append(num_tape)
+
+    def processing_fn(results):  # pylint: disable=too-many-branches
+        multi_measurements = len(tape.measurements) > 1
+        multi_params = len(tape.trainable_params) > 1
+
+        final_res = [
+            [qml.math.convert_like(2 * coeff * r, r) for r in res]
+            if isinstance(res, tuple)
+            else qml.math.convert_like(2 * coeff * res, res)
+            for coeff, res in zip(coeffs, results)
+        ]
+
+        # Post process for probs
+        if measurements_probs:
+            projector = np.array([1, -1])
+            if multi_measurements:
+                projector = qml.math.convert_like(projector, final_res[0][0])
+            else:
+                projector = qml.math.convert_like(projector, final_res[0])
+            for idx, res in enumerate(final_res):
+                if multi_measurements:
+                    for prob_idx in measurements_probs:
+                        num_wires_probs = len(tape.measurements[prob_idx].wires)
+                        res_reshaped = qml.math.reshape(res[prob_idx], (2**num_wires_probs, 2))
+                        final_res[idx][prob_idx] = qml.math.tensordot(
+                            res_reshaped, projector, axes=[[1], [0]]
+                        )
+                else:
+                    prob_idx = measurements_probs[0]
+                    num_wires_probs = len(tape.measurements[prob_idx].wires)
+                    res = qml.math.reshape(res, (2**num_wires_probs, 2))
+                    final_res[idx] = qml.math.tensordot(res, projector, axes=[[1], [0]])
+        grads = []
+
+        for idx, num_tape in enumerate(gradient_data):
+            if num_tape == 0:
+                grads.append(qml.math.zeros(()))
+            elif num_tape == 1:
+                grads.append(final_res[idx])
+            else:
+                axis = None if not multi_measurements else 0
+                grads.append(qml.math.sum(final_res[idx : idx + num_tape], axis=axis))
+
+        if not multi_measurements and not multi_params:
+            return grads[0]
+
+        if not (multi_params and multi_measurements):
+            if multi_measurements:
+                return tuple(grads[0])
+            return tuple(grads)
+
+        # Reordering to match the right shape for multiple measurements
+        grads_reorder = [[0] * len(tape.trainable_params) for _ in range(len(tape.measurements))]
+
+        for i in range(len(tape.measurements)):
+            for j in range(len(tape.trainable_params)):
+                grads_reorder[i][j] = grads[j][i]
+
+        grads_tuple = tuple(tuple(elem) for elem in grads_reorder)
+
+        return grads_tuple
+
+    return g_tapes, processing_fn
+
+
+def _get_generators(trainable_op):
```

Comment:
if we want to add something new here, it should also go in the list in tape_expand.py, right? maybe just put a note here telling devs to add to that list as well if they add it here (unless i'm misinterpreting this)

Reply:
Yes that's correct let me add a note!